plugin/core/registry.py

```diff
@@ -97,10 +97,9 @@ def _sessions_for_view_and_window(view: sublime.View, window: Optional[sublime.W
 
     manager = windows.lookup(window)
     scope_configs = manager._configs.scope_configs(view, point)
-    for config in scope_configs:
-        session = manager.get_session(config.name, file_path)
-        if session:
-            yield session
+    sessions = (manager.get_session(config.name, file_path) for config in scope_configs)
+    ready_sessions = (session for session in sessions if session and session.state == ClientStates.READY)
+    return ready_sessions
```

Comment:
I'm fine with removing the init lock, but will this not re-introduce the initialize race in the event listeners?

e.g. if the colorProvider feature initializes too early, it will get 0 sessions back from `get_sessions_for_view`. Same holds for the completionProvider, documentHighlightProvider. These event listeners will then never work again (they'll work once the user closes and opens the view again).

Reply:
Instead of this pull-model with `get_sessions_for_view`, I propose we remove that function entirely and go for a push-model in the event listeners. Something like this:

```python
class CompletionListener(LSPEventListener):

    def on_session_initialized(self, session: Session) -> None:
        # do bookkeeping for this session, keep it around to ask for completions
        pass

    def on_session_shutdown(self, session: Session) -> None:
        # do bookkeeping
        pass
```

In fact I'm already running a prototype of this idea locally, but it has way too many unrelated changes ;)

Reply:
I had a go at some kind of session lifecycle management for the vieweventlisteners today's but having to revoke sessions and handling multiple sessions per view made my attempt complicated quickly! 

I think even more benefits can be realised by refactoring to a single ViewEventListener per view (currently so many instances are created per view doing exactly the same checks!), and creating a new api for features that provides certain guarantees (a session, with requested capability, a valid file path, etc).

Reply:
The initialisation races were sort of a problem, but colorProvider had a 1-second retry hack and the other two examples need early user interaction before the server initialises for them to fail.

Reply:
> I think even more benefits can be realised by refactoring to a single ViewEventListener per view (currently so many instances are created per view doing exactly the same checks!)

Agreed. That's exactly what I ended up with in my prototype too :)

> and creating a new api for features that provides certain guarantees (a session, with requested capability, a valid file path, etc).

Hmm. I ended up with a `SessionView` object. Basically just a class that has state per session per view. If there are N sessions a view event listener would maintain N session views. So the total number of session views would be about ~ N * M, with M being the number of views.

This SessionView object solves the multi-session problem too ("how do we handle multiple sessions per view nicely?"), but like you mentioned there's not really a canonical way to handle, for instance, signature help from multiple sessions

I might push my branch soon so anyone can take a peek.

Reply:
Yes push the branch!